src/rod_mechanics/softRobots.cpp

```diff
@@ -57,6 +57,71 @@ void softRobots::applyInitialVelocities(int limb_idx, const vector<Vector3d> &ve
     }
 }
 
+void softRobots::applyPositionBC(const MatrixXd &positions) {
+    // Positions: the first col is limb_idx, second col is node_idx, third col is x, fourth col is y, fifth col is z 
+    if (positions.cols() != 5) {
+        throw runtime_error("The positions matrix should have 5 columns!");
+    }
+
+    for (int i = 0; i < positions.rows(); i++) {
+        int limb_idx = positions(i, 0);
+        int node_idx = positions(i, 1);
+        if (limb_idx >= limbs.size() || node_idx >= limbs[limb_idx]->nv) {
+            throw runtime_error("Invalid limb_idx or node_idx given!");
+        }
+        
+        Vector3d pos = positions.row(i).segment(2, 3).transpose();
+        limbs[limb_idx]->x.segment(4*node_idx, 3) += pos;
+    }
+}
+
+void softRobots::applyTwistBC(const MatrixXd &twists) {
+    // Twists: the first col is limb_idx, second col is node_idx, third col is theta 
+    if (twists.cols() != 3) {
+        throw runtime_error("The twist input matrix should have 3 columns!");
+    }
+
+    for (int i = 0; i < twists.rows(); i++) {
+        // the first col is limb_idx, second col is node_idx, third col is x, fourth col is y, fifth col is z 
+        int limb_idx = twists(i, 0);
+        int edge_idx = twists(i, 1);
+        if (limb_idx >= limbs.size() || edge_idx >= limbs[limb_idx]->ne) {
+            throw runtime_error("Invalid limb_idx or edge_idx given!");
+        }
+        
+        double theta = twists(i, 2);
+        limbs[limb_idx]->x(4*edge_idx + 3) += theta;
+    }
+}
+
+void softRobots::applyCurvatureBC(const MatrixXd &Curvatures) {
```

Comment:
Change datatype to `Matrix<double, Dynamic, 4>`. Helps users know right away the shape of the matrix.